libgadget/treewalk.c

```diff
@@ -281,47 +275,74 @@ cmpint(const void *a, const void *b)
 #endif
 
 static void
-treewalk_build_queue(TreeWalk * tw, int * active_set, int size) {
+treewalk_init_evaluated(int * active_set, int size)
+{
+    #pragma omp parallel for
+    for(int i=0; i < size; i++)
+    {
+        const int p_i = active_set ? active_set[i] : i;
+        P[p_i].Evaluated = 0;
+    }
+}
+
+static void
+treewalk_build_queue(TreeWalk * tw, int * active_set, const int size, int may_have_garbage) {
+    int i;
+
+    if(!tw->haswork && !may_have_garbage)
+    {
+        tw->WorkSetSize = size;
+        tw->WorkSet = active_set;
+        tw->work_set_allocated = 0;
+        return;
+    }
+
+    tw->WorkSet = mymalloc("ActiveQueue", size * sizeof(int));
+    tw->work_set_allocated = 1;
+
     int * queue = tw->WorkSet;
-    int k = 0;
-    if(active_set == NULL) {
-        int i;
-        #pragma omp parallel for
-        for(i = 0; i < PartManager->NumPart; i++) {
-            /* Skip the garbage particles */
-            if(P[i].IsGarbage) continue;
-            if(!tw->haswork(i, tw))
-                continue;
-            const int lock = atomic_fetch_and_add(&k, 1);
-            queue[lock] = i;
-        }
-    } else {
-        int i;
-        #pragma omp parallel for
-        for(i=0; i < NumActiveParticle; i++)
-        {
-            const int p_i = ActiveParticle[i];
+    int nqueue = 0;
 
-            /* Skip the garbage particles */
-            if(P[p_i].IsGarbage) continue;
+    /*We want a lockless algorithm which preserves the ordering of the particle list.*/
+    int nqthr[All.NumThreads];
+    memset(nqthr, 0, All.NumThreads*sizeof(int));
+    int * thrqueue = mymalloc("threadedqueue", size * sizeof(int) * All.NumThreads);
+
+    /* We enforce schedule static to ensure that each thread executes on contiguous particles.*/
+    #pragma omp parallel for schedule(static)
+    for(i=0; i < size; i++)
+    {
+        const int tid = omp_get_thread_num();
+        /*Use raw particle number if active_set is null, otherwise use active_set*/
+        const int p_i = active_set ? active_set[i] : i;
+
+        /* Skip the garbage particles */
+        if(P[p_i].IsGarbage) continue;
+
+        if(tw->haswork && !tw->haswork(p_i, tw))
+            continue;
+        thrqueue[size * tid + nqthr[tid]] = p_i;
+        nqthr[tid]++;
+    }
+    /*Merge step for the queue.*/
+    nqueue = 0;
```

Comment:
This showed up twice. Would a function like this be useful?

```
gadget_compact_arrays (void * dest, void * srcs[], size_t sizes[], size_t elsize, int narrays)
```

Reply:
Done - although I made it for integers only because I was scared of casting int ** to void ** 
It doesn't save any lines of code but I think it is clearer.